Module1 slides1 (#9)
UBC-MDS/exploratory-data-viz: chapters/en/slides/module1/module1_01.Rmd

for set_name, df in anscombe.groupby('Set'):
    df = df.drop(columns='Set')
    df.columns = pd.MultiIndex.from_tuples([(set_name, 'X'), (set_name, 'Y')])
    print(
        df
        .to_html(index=False)
        .replace('class="dataframe"', 'style="width:70px; float:left; margin-right:40px"')
        .replace('<thead>',  '<thead style="text-align: center">'))
```

Notes:
Think about it,
numbers have only been around for about 5,000 years.
Visual systems on the other hand,
have undergone refinement during 500,000,000 years of evolution.
While we need to train ourselves to recognize structure in numerical data,
we have evolved to instinctively recognize visual patterns
and to accurately judge properties such as colors and distances between objects.
Drawing upon this complex build-in apparatus,
we can arrive at conclusion much faster by glancing at visual representations of data
rather than numerical representations.

For example, have a look at the
four sets of numbers in the table on the slide.
This is a slightly modified version of the original,
[which was put together by statistician Francis Anscombe in the
70s.](https://en.wikipedia.org/wiki/Anscombe%27s_quartet)

> Can you see how the differences in the general trends between these four sets of numbers?

Reviewer: ```suggestion
Can you see the differences in the general trends between these four sets of numbers?
```

### C is the only set with a different mean and standard deviation

```{python echo=FALSE, results='asis'}
for set_name, df in anscombe.groupby('Set'):
    df = df.drop(columns='Set')
    df.columns = pd.MultiIndex.from_tuples([(set_name, 'X'), (set_name, 'Y')])
    print(
        df
        .agg(['mean', 'std']).round(2).reset_index().rename(columns={'index': ''})
        .to_html(index=False)
        .replace('class="dataframe"', 'style="width:70px; float:left; margin-right:40px"')
        .replace('<thead>',  '<thead style="text-align: center">'))
```

Notes:
Summaries,
such as the mean and standard deviation,
are helpful statistical tools
that are often useful for detecting difference between datasets.
However,
since they collapse the data into just a few numbers,
statistical summaries can't tell the whole story about the data
and there can be important differences between datasets
that summaries fail to reveal.
Here,
the mean and standard deviation indicate that set C
is slightly different from the other sets of data
in terms of the centre of the sample distribution and the spread of that distribution,

> while the remaining three sets of data have a similar centre and spread.

Reviewer: Can you highlight the differences in C? With a red circle around the number they should first pay attention to? Or shade the background of that cell? Anything that is not too difficult to do that draws attention quickly to it would help.

Author: Is bolding too little? It is the easiest, but I might be able to add a color ful span, let's see 
![image](https://user-images.githubusercontent.com/4560057/101963076-aaca2200-3bc2-11eb-9160-ee8a5d015bda.png)

Author: I think I should also be able to get this working if you think it is better 
![image](https://user-images.githubusercontent.com/4560057/101963417-81f65c80-3bc3-11eb-89c7-07dd241a1449.png)

Author: Ah this is my favorite, mathcing the color to the color theme 
![image](https://user-images.githubusercontent.com/4560057/101963558-dac5f500-3bc3-11eb-85f7-ee6ce4b98b7f.png)

> ### Can you spot the differences between these four sets of numbers?

Reviewer: This question needs rephrasing - we can spot many differences... the exact numbers are different! Maybe the question should be "Can you see differences in trends or patterns between these four sets of numbers?"

```{python echo=FALSE, results='asis'}
# This is the code that generates the table below,
# and setting results to `'asis'` ensures it is rendered as HTML
# instead of being wrapped in an output code block
import altair as alt
from vega_datasets import data

anscombe = data.anscombe()#.query('Series != "IV"')
anscombe['Set'] = anscombe['Series'].map({'I': 'A', 'II': 'B', 'III': 'C', 'IV': 'D'})
anscombe = anscombe.drop(columns='Series')
anscombe.loc[anscombe['Y'] == anscombe['Y'].max(), 'Y'] = 8.5

for set_name, df in anscombe.groupby('Set'):
    df = df.drop(columns='Set')
    df.columns = pd.MultiIndex.from_tuples([(set_name, 'X'), (set_name, 'Y')])
    print(
        df
        .to_html(index=False)
        .replace('class="dataframe"', 'style="width:70px; float:left; margin-right:40px"')
        .replace('<thead>',  '<thead style="text-align: center">'))
```

Notes:
Think about it,
numbers have only been around for about 5,000 years.
Visual systems on the other hand,
have undergone refinement for over 500,000,000 years

> by training with a lethal, evolutionary cost function.

Reviewer: Non-bio majors are going to not understand "training with a lethal, evolutionary cost function" can we simply/reduce jargon here? I like it, but for a general audience this could confuse.

Author: I was thinking "cost function" would be OK given that they just had the ML course, but I mayeb not all students have taken that? I rephrased.

Reviewer: No, not all students will have taken that. The only pre-req for this course is the programming course.

While we need to train ourselves to recognize structure in numerical data,
we have evolved to instinctively recognize visual patterns
and to accurately judge properties such as colors and distances between objects.
Drawing upon this complex build-in apparatus,
we can arrive at conclusion much faster by glancing at visual representations of data
rather than numerical representations.

For example, have a look at the
four sets of numbers in the table on the slide.
This is a slightly modified version of the original,
[which was put together by statistician Francis Anscombe in the
70s.](https://en.wikipedia.org/wiki/Anscombe%27s_quartet)
Can you see how the sets are different from each other

> by just glancing at the table?

Reviewer: again, I would use the phrasing "Can you see the differences in trends or patterns between these four sets of numbers?"

---

## Although summary statistics are often useful, they don't tell the whole story

> ### Set C appears to be slightly different, but are the other sets actually identical?

Reviewer: We know that the other sets are not actually identical - we just viewed the raw numbers. Can we highlight this here somehow in a useful way? Maybe "The summary statistics, mean and standard deviation, indicate C is very/quite (find the right level of adjective for here) different from the other sets of data in terms of the centre of the sample distribution and the spread of that distribution, and that the remaining 3 sets of data have a similar centre and spread." Or something like that?

```{python echo=FALSE, results='asis'}
for set_name, df in anscombe.groupby('Set'):
    df = df.drop(columns='Set')
    df.columns = pd.MultiIndex.from_tuples([(set_name, 'X'), (set_name, 'Y')])
    print(
        df
        .agg(['mean', 'std']).round(2).reset_index().rename(columns={'index': ''})
        .to_html(index=False)
        .replace('class="dataframe"', 'style="width:70px; float:left; margin-right:40px"')
        .replace('<thead>',  '<thead style="text-align: center">'))
```

Notes:
Summaries,
such as the mean and standard deviation,
are helpful statistical tools
that are often useful for detecting difference between datasets.
However,
since they collapse the data into just a few numbers,
statistical summaries can't tell the whole story about the data
and there can be important differences between datasets
that summaries fail to reveal.

---

## Plotting the data immediately reveals the differences to the human eye

> ### There are clear patterns in these datasets that we could not detect from looking at the numbers

Reviewer: ```suggestion
### There are clear patterns in these datasets that we could not detect from only looking at some sample statistics
```

Author: I rewrote slightly

            'legend': {'symbolSize': 14, 'titleFontSize': 14, 'labelFontSize': 14},
            'axis': {'titleFontSize': 14, 'labelFontSize': 12},
            'header': {'titleFontSize': 16, 'labelFontSize': 14},
            'encoding': {'x': {'scale': {'zero': False}}}}}
alt.themes.register('bigger_font', bigger_font)
alt.themes.enable('bigger_font')

(alt.Chart(anscombe).mark_circle(size=80, opacity=1).encode(
    x='X',
    y='Y')
 .properties(width=300, height=200)
 .facet('Set', columns=2)
 .configure_axis(grid=True)
 .configure_header(title=None)
 .save(path + 'modified-anscombes.svg'))
```

<img src="/module1/modified-anscombes.svg" alt="" width="65%"></img>

Notes:
Humans are not good at detecting patterns in raw numbers,
and we don't have good intuition about how different distributions of data
can contribute to identical statistical summaries.
But guess what we excel at?
Detecting visual patterns!
It is immediately clear to us how these sets of numbers differ
once they are shown as graphical elements
instead of textual objects.
This is one of the main reasons why data visualization
is such a powerful tool for data exploration and communication.

> 

Reviewer: Maybe spend some time talking about how these sample distributions differ, and how that could matter?
